content/post/fewer-parameters-means-fewer-bugs.md

```diff
@@ -0,0 +1,244 @@
+
+This practice is also known as [Duck Typing][1]. If it walks like a duck and it quacks like a duck, then it must be a duck. In duck typing, an object's suitability is determined by the presence of certain methods and properties, rather than the type of the object itself. In the Java case, you can use the interface.
+
+Now a new requirement is coming, we need to add a new method called `cover`, and it also has author first name and last name inside. We can follow the pattern we just made. Let's compare before and after the refactoring of the book class.
+#### Old pattern without refactoring
+```rb
+def summary(first_name, last_name)
+  "#{title} - #{last_name}, #{first_name} (#{year_published})"
+end
+ 
+def cover(first_name, last_name)
+  "#{title}\n\n\n#{last_name}, #{first_name}"
+end
+```
+#### After refactor
+```rb
+def summary(author)
+  "#{title} - #{author.full_name} (#{year_published})"
+end
+ 
+def cover(author)
+  "#{title}\n\n\n#{author.full_name}"
+end
+```
+The cover method generates a string that has the book title and author's full name inside. After the refactoring, we can reuse the author's full name method again. Once we build up a good pattern, the new code can follow it easier and keep our project clean. But if you didn't pay attention, the code would quickly become harder to read, develop technical debt and become a developer's worst nightmare.
+
+You may also notice that both `summary` and `cover` need the same parameter, author, it's the smell of refactoring. The author information should be one of the attributes of the book object, instead of passing as a parameter when we call the book method. So we can refactor the book class again.
+
+```rb
+class Book
+  attr_reader :title, :year_published, :author
+  
+  def initialize(title, year_published, author)
+    @title = title
+    @year_published = year_published
+    @author = author
+  end
+ 
+  def summary
+    "#{title} - #{author.full_name} (#{year_published})"
+  end
+  
+  def cover
+    "#{title}\n\n\n#{author.full_name}"
+  end
+end
+  
+bater = Author.new("Bater", "Chen")
+book = Book.new("Ruby is the best", "2020", bater)
+book.summary
+=> "Ruby is the best - Chen, Bater (2020)"
+book.cover
+=> "Ruby is the best\n\n\nChen, Bater"
+```
+Both `summary` and `cover` methods' output remain the same but now we can call it freely without any parameter once the book object is prepared. In Ruby, we can skip parentheses when there is no parameter.
+
+This example doesn't encourage you to move all the parameters into the object attributes, it depends on the domain know-how. We can do it because it makes sense for a book object to having an author as an attribute in this case. It may get worse if we try to put the wrong data as the object attribute.
```

Comment:
```suggestion
This example doesn't encourage you to move all the parameters into the object attributes --- doing this depends on the domain knowledge. We can do this here because it makes sense for a `Book` object to having an `author` as an attribute in this case. But in other cases, the code may actually get worse if we try to put the wrong data in an object attribute.
```